### PR TITLE
Refactor and optimize shader rendering and compositing

### PR DESCRIPTION
**Pull request by xian**
- Output pixels are now passed to `Shader.render()`.
- `CompositorShader` uses one fewer extra pixel buffers for sub-shader rendering.
- Added tests for `CompositorShader`.

### CODE REVIEW

File: src/commonTest/kotlin/baaahs/shaders/PixelShaderTest.kt [hunk 1]
```diff
@@ -1,57 +1,38 @@
...
-
-        val reader = ByteArrayReader(bytes)
-        expect(srcShader.id.ordinal.toByte()) { reader.readByte() }
-
-        val dstShader: Shader<T> = srcShader.id.reader.parse(reader) as Shader<T>
-        val dstBuf = dstShader.createBuffer(surface)
-        dstBuf.read(reader)
-        return dstBuf
+    @Test
+    fun whenMorePixels_shouldRepeat() {
```
Comment (hassoncs) [inline src/commonTest/kotlin/baaahs/shaders/PixelShaderTest.kt]: 🆒💯

File: src/commonMain/kotlin/baaahs/shaders/CompositorShader.kt [hunk 2]
```diff
@@ -99,8 +105,12 @@ class CompositorShader(val aShader: Shader<*>, val bShader: Shader<*>) :
 }
 
 enum class CompositingMode {
-    OVERLAY { override fun composite(src: Color, dest: Color) = src },
-    ADD { override fun composite(src: Color, dest: Color) = dest.plus(src) };
+    OVERLAY {
```
[the remark this answers is not in the document]
Comment (xian) [reply src/commonMain/kotlin/baaahs/shaders/CompositorShader.kt]: Awesome point. I've been thinking of the Porter-Duff modes as the canonical ones here, and they don't exactly map PSD, which is certainly way more well-known. The PD version is called something like `SRC_OVER` which doesn't quite spring from one's tongue either. Maybe we should just go with the PSD mode names?

[the remark this answers is not in the document]
Comment (xian) [reply src/commonMain/kotlin/baaahs/shaders/CompositorShader.kt]: `NORMAL` seems pretty sensible to me, I'll change it tomorrow. :-)

[the remark this answers is not in the document]
Comment (xian) [reply src/commonMain/kotlin/baaahs/shaders/CompositorShader.kt]: Fixed!

**Pull request merged**